From d4afc34dafa7e33f04d43e0b36d9ed9192640985 Mon Sep 17 00:00:00 2001
From: David Wagner <david.wagner@intel.com>
Date: Wed, 7 Aug 2013 17:03:26 +0200
Subject: AudioPolicy: Add an "idle" strategy

BZ: 137361
[CLEAN]

This new strategy applies when no other does (hence the name "idle").
It elects a wired accessory if any is available or falls back to the
builtin speaker.

This allows the accessory change information to be forwarded even if there
is no active stream.  That is necessary for external routes that are not
represented by stream types (such as "media") in the audio policy.

Category: feature differentiation
Domain: AudioComms-VoiceTrigger
Origin: internal
Upstream-Candidate: yes
Change-Id: I415d1a5d113a66c00b4a168093bd77ec05520df3
Orig-Change-Id: I8c16485344b515a850b06e7c104cbba715354bba
Signed-off-by: David Wagner <david.wagner@intel.com>
---
 audio/AudioPolicyManagerBase.cpp                 |   35 +++++++++++++++++++++-
 include/hardware_legacy/AudioPolicyManagerBase.h |    1 +
 2 files changed, 35 insertions(+), 1 deletions(-)

diff --git a/audio/AudioPolicyManagerBase.cpp b/audio/AudioPolicyManagerBase.cpp
index 137b166..269c274 100644
--- a/audio/AudioPolicyManagerBase.cpp
+++ b/audio/AudioPolicyManagerBase.cpp
@@ -2221,6 +2221,7 @@ void AudioPolicyManagerBase::checkOutputForAllStrategies()
     checkOutputForStrategy(STRATEGY_SONIFICATION_LOCAL);
     checkOutputForStrategy(STRATEGY_MEDIA);
     checkOutputForStrategy(STRATEGY_DTMF);
+    checkOutputForStrategy(STRATEGY_IDLE);
 }
 
 audio_io_handle_t AudioPolicyManagerBase::getA2dpOutput() const
@@ -2319,6 +2320,8 @@ audio_devices_t AudioPolicyManagerBase::getNewDevice(audio_io_handle_t output, b
         device = getDeviceForStrategy(STRATEGY_MEDIA, fromCache);
     } else if (outputDesc->isStrategyActive(STRATEGY_DTMF)) {
         device = getDeviceForStrategy(STRATEGY_DTMF, fromCache);
+    } else {
+        device = getDeviceForStrategy(STRATEGY_IDLE, fromCache);
     }
 
     ALOGV("getNewDevice() selected device %x", device);
@@ -2590,7 +2593,37 @@ audio_devices_t AudioPolicyManagerBase::getDeviceForStrategy(routing_strategy st
             ALOGE("getDeviceForStrategy() no device found for STRATEGY_MEDIA");
         }
         } break;
-
+    case STRATEGY_IDLE: {
+        if (device == AUDIO_DEVICE_NONE) {
+            device = mAvailableOutputDevices & AUDIO_DEVICE_OUT_WIRED_HEADPHONE;
+        }
+        if (device == AUDIO_DEVICE_NONE) {
+            device = mAvailableOutputDevices & AUDIO_DEVICE_OUT_WIRED_HEADSET;
+        }
+        if (device == AUDIO_DEVICE_NONE) {
+            device = mAvailableOutputDevices & AUDIO_DEVICE_OUT_USB_ACCESSORY;
+        }
+        if (device == AUDIO_DEVICE_NONE) {
+            device = mAvailableOutputDevices & AUDIO_DEVICE_OUT_USB_DEVICE;
+        }
+        if (device == AUDIO_DEVICE_NONE) {
+            device = mAvailableOutputDevices & AudioSystem::DEVICE_OUT_DGTL_DOCK_HEADSET;
+        }
+        if (device == AUDIO_DEVICE_NONE) {
+            device = mAvailableOutputDevices & AudioSystem::DEVICE_OUT_AUX_DIGITAL;
+        }
+        if (device == AUDIO_DEVICE_NONE) {
+            device = mAvailableOutputDevices & AUDIO_DEVICE_OUT_REMOTE_SUBMIX;
+        }
+        if ((device == AUDIO_DEVICE_NONE) &&
+                (mForceUse[AudioSystem::FOR_DOCK] == AudioSystem::FORCE_ANALOG_DOCK)) {
+            device = mAvailableOutputDevices & AUDIO_DEVICE_OUT_ANLG_DOCK_HEADSET;
+        }
+        if (device == AUDIO_DEVICE_NONE) {
+            device = mAvailableOutputDevices & AUDIO_DEVICE_OUT_SPEAKER;
+        }
+        }
+        break;
     default:
         ALOGW("getDeviceForStrategy() unknown strategy: %d", strategy);
         break;
diff --git a/include/hardware_legacy/AudioPolicyManagerBase.h b/include/hardware_legacy/AudioPolicyManagerBase.h
index 47256d6..293269e 100644
--- a/include/hardware_legacy/AudioPolicyManagerBase.h
+++ b/include/hardware_legacy/AudioPolicyManagerBase.h
@@ -158,6 +158,7 @@ protected:
             STRATEGY_SONIFICATION_LOCAL,
             STRATEGY_DTMF,
             STRATEGY_ENFORCED_AUDIBLE,
+            STRATEGY_IDLE,
             NUM_STRATEGIES
         };
 
-- 
1.7.4.1

